// @mostly-good-metrics/javascript/dist/cjs/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MostlyGoodMetrics = void 0;
const logger_1 = require("./logger");
const network_1 = require("./network");
const storage_1 = require("./storage");
const types_1 = require("./types");
const utils_1 = require("./utils");
const FLUSH_DELAY_MS = 100; // Delay between batch sends
/**
 * Main client for MostlyGoodMetrics.
 * Use the static `configure` method to initialize, then use static methods or the instance.
 */
class MostlyGoodMetrics {
    /**
     * Private constructor - use `configure` to create an instance.
     */
    constructor(config) {
        this.flushTimer = null;
        this.isFlushingInternal = false;
        this.lifecycleSetup = false;
        this.handleVisibilityChange = () => {
            if (document.hidden) {
                // App backgrounded
                this.track(types_1.SystemEvents.APP_BACKGROUNDED);
                void this.flush(); // Flush when going to background
            }
            else {
                // App foregrounded
                this.track(types_1.SystemEvents.APP_OPENED);
            }
        };
        this.handleBeforeUnload = () => {
            // Best-effort flush using sendBeacon if available
            this.flushWithBeacon();
        };
        this.handlePageHide = () => {
            // Best-effort flush using sendBeacon if available
            this.flushWithBeacon();
        };
        this.config = (0, utils_1.resolveConfiguration)(config);
        this.sessionIdValue = (0, utils_1.generateUUID)();
        // Set up logging
        (0, logger_1.setDebugLogging)(this.config.enableDebugLogging);
        // Initialize storage
        this.storage = this.config.storage ?? (0, storage_1.createDefaultStorage)(this.config.maxStoredEvents);
        // Initialize network client
        this.networkClient = this.config.networkClient ?? (0, network_1.createDefaultNetworkClient)();
        logger_1.logger.info(`MostlyGoodMetrics initialized with environment: ${this.config.environment}`);
        // Start auto-flush timer
        this.startFlushTimer();
        // Set up lifecycle tracking
        if (this.config.trackAppLifecycleEvents) {
            this.setupLifecycleTracking();
        }
    }
    /**
     * Configure and initialize the SDK.
     * Returns the singleton instance.
     */
    static configure(config) {
        if (!config.apiKey) {
            throw new Error('API key is required');
        }
        if (MostlyGoodMetrics.instance) {
            logger_1.logger.warn('MostlyGoodMetrics.configure called multiple times. Using existing instance.');
            return MostlyGoodMetrics.instance;
        }
        MostlyGoodMetrics.instance = new MostlyGoodMetrics(config);
        return MostlyGoodMetrics.instance;
    }
    /**
     * Get the shared instance, or null if not configured.
     */
    static get shared() {
        return MostlyGoodMetrics.instance;
    }
    /**
     * Check if the SDK has been configured.
     */
    static get isConfigured() {
        return MostlyGoodMetrics.instance !== null;
    }
    /**
     * Reset the SDK (primarily for testing).
     */
    static reset() {
        if (MostlyGoodMetrics.instance) {
            MostlyGoodMetrics.instance.destroy();
            MostlyGoodMetrics.instance = null;
        }
    }
    // ============================================================
    // Static convenience methods (delegate to shared instance)
    // ============================================================
    /**
     * Track an event with the given name and optional properties.
     */
    static track(name, properties) {
        MostlyGoodMetrics.instance?.track(name, properties);
    }
    /**
     * Identify the current user.
     */
    static identify(userId) {
        MostlyGoodMetrics.instance?.identify(userId);
    }
    /**
     * Reset user identity.
     */
    static resetIdentity() {
        MostlyGoodMetrics.instance?.resetIdentity();
    }
    /**
     * Flush pending events to the server.
     */
    static flush() {
        return MostlyGoodMetrics.instance?.flush() ?? Promise.resolve();
    }
    /**
     * Start a new session.
     */
    static startNewSession() {
        MostlyGoodMetrics.instance?.startNewSession();
    }
    /**
     * Clear all pending events.
     */
    static clearPendingEvents() {
        return MostlyGoodMetrics.instance?.clearPendingEvents() ?? Promise.resolve();
    }
    /**
     * Get the count of pending events.
     */
    static getPendingEventCount() {
        return MostlyGoodMetrics.instance?.getPendingEventCount() ?? Promise.resolve(0);
    }
    /**
     * Set a single super property that will be included with every event.
     */
    static setSuperProperty(key, value) {
        MostlyGoodMetrics.instance?.setSuperProperty(key, value);
    }
    /**
     * Set multiple super properties at once.
     */
    static setSuperProperties(properties) {
        MostlyGoodMetrics.instance?.setSuperProperties(properties);
    }
    /**
     * Remove a single super property.
     */
    static removeSuperProperty(key) {
        MostlyGoodMetrics.instance?.removeSuperProperty(key);
    }
    /**
     * Clear all super properties.
     */
    static clearSuperProperties() {
        MostlyGoodMetrics.instance?.clearSuperProperties();
    }
    /**
     * Get all current super properties.
     */
    static getSuperProperties() {
        return MostlyGoodMetrics.instance?.getSuperProperties() ?? {};
    }
    // ============================================================
    // Instance properties
    // ============================================================
    /**
     * Get the current user ID.
     */
    get userId() {
        return storage_1.persistence.getUserId();
    }
    /**
     * Get the current session ID.
     */
    get sessionId() {
        return this.sessionIdValue;
    }
    /**
     * Check if a flush operation is in progress.
     */
    get isFlushing() {
        return this.isFlushingInternal;
    }
    /**
     * Get the resolved configuration.
     */
    get configuration() {
        return { ...this.config };
    }
    // ============================================================
    // Instance methods
    // ============================================================
    /**
     * Track an event with the given name and optional properties.
     */
    track(name, properties) {
        try {
            (0, utils_1.validateEventName)(name);
        }
        catch (e) {
            logger_1.logger.error(`Invalid event name: ${name}`, e);
            return;
        }
        const sanitizedProperties = (0, utils_1.sanitizeProperties)(properties);
        const superProperties = storage_1.persistence.getSuperProperties();
        // Merge properties: super properties < event properties < system properties
        // Event properties override super properties, system properties are always added
        const mergedProperties = {
            ...superProperties,
            ...sanitizedProperties,
            [types_1.SystemProperties.DEVICE_TYPE]: (0, utils_1.detectDeviceType)(),
            [types_1.SystemProperties.DEVICE_MODEL]: (0, utils_1.getDeviceModel)(),
            [types_1.SystemProperties.SDK]: this.config.sdk,
        };
        const event = {
            name,
            client_event_id: (0, utils_1.generateUUID)(),
            timestamp: (0, utils_1.getISOTimestamp)(),
            userId: this.userId ?? undefined,
            sessionId: this.sessionIdValue,
            platform: this.config.platform,
            appVersion: this.config.appVersion || undefined,
            osVersion: this.config.osVersion || (0, utils_1.getOSVersion)() || undefined,
            environment: this.config.environment,
            locale: (0, utils_1.getLocale)(),
            timezone: (0, utils_1.getTimezone)(),
            properties: Object.keys(mergedProperties).length > 0 ? mergedProperties : undefined,
        };
        logger_1.logger.debug(`Tracking event: ${name}`, event);
        // Store event asynchronously
        this.storage.store(event).catch((e) => {
            logger_1.logger.error('Failed to store event', e);
        });
        // Check if we should flush due to batch size
        void this.checkBatchSize();
    }
    /**
     * Identify the current user.
     */
    identify(userId) {
        if (!userId) {
            logger_1.logger.warn('identify called with empty userId');
            return;
        }
        logger_1.logger.debug(`Identifying user: ${userId}`);
        storage_1.persistence.setUserId(userId);
    }
    /**
     * Reset user identity.
     */
    resetIdentity() {
        logger_1.logger.debug('Resetting user identity');
        storage_1.persistence.setUserId(null);
    }
    /**
     * Start a new session.
     */
    startNewSession() {
        this.sessionIdValue = (0, utils_1.generateUUID)();
        logger_1.logger.debug(`Started new session: ${this.sessionIdValue}`);
    }
    /**
     * Flush pending events to the server.
     */
    async flush() {
        if (this.isFlushingInternal) {
            logger_1.logger.debug('Flush already in progress');
            return;
        }
        this.isFlushingInternal = true;
        logger_1.logger.debug('Starting flush');
        try {
            await this.performFlush();
        }
        finally {
            this.isFlushingInternal = false;
        }
    }
    /**
     * Clear all pending events.
     */
    async clearPendingEvents() {
        logger_1.logger.debug('Clearing all pending events');
        await this.storage.clear();
    }
    /**
     * Get the count of pending events.
     */
    async getPendingEventCount() {
        return this.storage.eventCount();
    }
    /**
     * Set a single super property that will be included with every event.
     */
    setSuperProperty(key, value) {
        logger_1.logger.debug(`Setting super property: ${key}`);
        storage_1.persistence.setSuperProperty(key, value);
    }
    /**
     * Set multiple super properties at once.
     */
    setSuperProperties(properties) {
        logger_1.logger.debug(`Setting super properties: ${Object.keys(properties).join(', ')}`);
        storage_1.persistence.setSuperProperties(properties);
    }
    /**
     * Remove a single super property.
     */
    removeSuperProperty(key) {
        logger_1.logger.debug(`Removing super property: ${key}`);
        storage_1.persistence.removeSuperProperty(key);
    }
    /**
     * Clear all super properties.
     */
    clearSuperProperties() {
        logger_1.logger.debug('Clearing all super properties');
        storage_1.persistence.clearSuperProperties();
    }
    /**
     * Get all current super properties.
     */
    getSuperProperties() {
        return storage_1.persistence.getSuperProperties();
    }
    /**
     * Clean up resources (stop timers, etc.).
     */
    destroy() {
        this.stopFlushTimer();
        this.removeLifecycleListeners();
        logger_1.logger.debug('MostlyGoodMetrics instance destroyed');
    }
    // ============================================================
    // Private methods
    // ============================================================
    async checkBatchSize() {
        const count = await this.storage.eventCount();
        if (count >= this.config.maxBatchSize) {
            logger_1.logger.debug('Batch size threshold reached, triggering flush');
            void this.flush();
        }
    }
    async performFlush() {
        let hasMoreEvents = true;
        while (hasMoreEvents) {
            const eventCount = await this.storage.eventCount();
            if (eventCount === 0) {
                logger_1.logger.debug('No events to flush');
                break;
            }
            // Check rate limiting
            if (this.networkClient.isRateLimited()) {
                logger_1.logger.debug('Rate limited, skipping flush');
                break;
            }
            const events = await this.storage.fetchEvents(this.config.maxBatchSize);
            if (events.length === 0) {
                break;
            }
            const payload = this.buildPayload(events);
            const result = await this.networkClient.sendEvents(payload, this.config);
            if (result.success) {
                logger_1.logger.debug(`Successfully sent ${events.length} events`);
                await this.storage.removeEvents(events.length);
            }
            else {
                logger_1.logger.warn(`Failed to send events: ${result.error.message}`);
                // Call onError callback if configured
                if (this.config.onError) {
                    try {
                        this.config.onError(result.error);
                    }
                    catch (e) {
                        logger_1.logger.error('Error in onError callback', e);
                    }
                }
                if (!result.shouldRetry) {
                    // Drop events on non-retryable errors (4xx)
                    logger_1.logger.warn('Dropping events due to non-retryable error');
                    await this.storage.removeEvents(events.length);
                }
                else {
                    // Keep events for retry on retryable errors
                    hasMoreEvents = false;
                }
            }
            // Small delay between batches to avoid overwhelming the server
            await (0, utils_1.delay)(FLUSH_DELAY_MS);
        }
    }
    buildPayload(events) {
        const context = {
            platform: this.config.platform,
            appVersion: this.config.appVersion || undefined,
            osVersion: this.config.osVersion || (0, utils_1.getOSVersion)() || undefined,
            userId: this.userId ?? undefined,
            sessionId: this.sessionIdValue,
            environment: this.config.environment,
            locale: (0, utils_1.getLocale)(),
            timezone: (0, utils_1.getTimezone)(),
        };
        return { events, context };
    }
    startFlushTimer() {
        if (this.flushTimer) {
            return;
        }
        this.flushTimer = setInterval(() => {
            void this.flush();
        }, this.config.flushInterval * 1000);
        logger_1.logger.debug(`Started flush timer (${this.config.flushInterval}s interval)`);
    }
    stopFlushTimer() {
        if (this.flushTimer) {
            clearInterval(this.flushTimer);
            this.flushTimer = null;
            logger_1.logger.debug('Stopped flush timer');
        }
    }
    setupLifecycleTracking() {
        if (this.lifecycleSetup) {
            return;
        }
        if (typeof window === 'undefined' || typeof document === 'undefined') {
            logger_1.logger.debug('Not in browser environment, skipping lifecycle tracking');
            return;
        }
        this.lifecycleSetup = true;
        // Track app installed/updated
        this.trackInstallOrUpdate();
        // Track app opened
        this.trackAppOpened();
        // Track visibility changes (background/foreground)
        document.addEventListener('visibilitychange', this.handleVisibilityChange);
        // Flush on page unload
        window.addEventListener('beforeunload', this.handleBeforeUnload);
        window.addEventListener('pagehide', this.handlePageHide);
        logger_1.logger.debug('Lifecycle tracking enabled');
    }
    removeLifecycleListeners() {
        if (typeof document !== 'undefined') {
            document.removeEventListener('visibilitychange', this.handleVisibilityChange);
        }
        if (typeof window !== 'undefined') {
            window.removeEventListener('beforeunload', this.handleBeforeUnload);
            window.removeEventListener('pagehide', this.handlePageHide);
        }
    }
    trackInstallOrUpdate() {
        const currentVersion = this.config.appVersion;
        const previousVersion = storage_1.persistence.getAppVersion();
        if (!currentVersion) {
            // No version configured, skip install/update tracking
            return;
        }
        if (storage_1.persistence.isFirstLaunch()) {
            // First launch ever - track install
            this.track(types_1.SystemEvents.APP_INSTALLED, {
                [types_1.SystemProperties.VERSION]: currentVersion,
            });
            storage_1.persistence.setAppVersion(currentVersion);
        }
        else if (previousVersion && previousVersion !== currentVersion) {
            // Version changed - track update
            this.track(types_1.SystemEvents.APP_UPDATED, {
                [types_1.SystemProperties.VERSION]: currentVersion,
                [types_1.SystemProperties.PREVIOUS_VERSION]: previousVersion,
            });
            storage_1.persistence.setAppVersion(currentVersion);
        }
        else if (!previousVersion) {
            // First time with version tracking
            storage_1.persistence.setAppVersion(currentVersion);
        }
    }
    trackAppOpened() {
        this.track(types_1.SystemEvents.APP_OPENED);
    }
    flushWithBeacon() {
        // Use sendBeacon for reliable delivery during page unload
        if (typeof navigator === 'undefined' || !navigator.sendBeacon) {
            return;
        }
        // Note: This is a synchronous, best-effort send
        // We can't use async storage operations here, so we rely on
        // the regular flush mechanism for most events
        logger_1.logger.debug('Page unloading, attempting beacon flush');
    }
}
exports.MostlyGoodMetrics = MostlyGoodMetrics;
MostlyGoodMetrics.instance = null;
