// @mostly-good-metrics/javascript/dist/cjs/index.js
"use strict";
/**
 * MostlyGoodMetrics JavaScript SDK
 *
 * A lightweight, framework-agnostic analytics library for web applications.
 *
 * @example
 * ```typescript
 * import { MostlyGoodMetrics } from '@mostly-good-metrics/javascript';
 *
 * // Initialize the SDK
 * MostlyGoodMetrics.configure({
 *   apiKey: 'mgm_proj_your_api_key',
 *   environment: 'production',
 * });
 *
 * // Track events
 * MostlyGoodMetrics.track('button_clicked', {
 *   button_id: 'submit',
 *   page: '/checkout',
 * });
 *
 * // Identify users with profile data
 * MostlyGoodMetrics.identify('user_123', {
 *   email: 'user@example.com',
 *   name: 'Jane Doe',
 * });
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = exports.getDeviceModel = exports.getOSVersion = exports.detectDeviceType = exports.detectPlatform = exports.sanitizeProperties = exports.validateEventName = exports.isValidEventName = exports.getISOTimestamp = exports.generateUUID = exports.generateAnonymousId = exports.createDefaultNetworkClient = exports.FetchNetworkClient = exports.createDefaultStorage = exports.LocalStorageEventStorage = exports.InMemoryEventStorage = exports.EVENT_NAME_REGEX = exports.Constraints = exports.DefaultConfiguration = exports.SystemProperties = exports.SystemEvents = exports.MGMError = exports.MostlyGoodMetrics = void 0;
// Main client
var client_1 = require("./client");
Object.defineProperty(exports, "MostlyGoodMetrics", { enumerable: true, get: function () { return client_1.MostlyGoodMetrics; } });
// Error class
var types_1 = require("./types");
Object.defineProperty(exports, "MGMError", { enumerable: true, get: function () { return types_1.MGMError; } });
// Constants
var types_2 = require("./types");
Object.defineProperty(exports, "SystemEvents", { enumerable: true, get: function () { return types_2.SystemEvents; } });
Object.defineProperty(exports, "SystemProperties", { enumerable: true, get: function () { return types_2.SystemProperties; } });
Object.defineProperty(exports, "DefaultConfiguration", { enumerable: true, get: function () { return types_2.DefaultConfiguration; } });
Object.defineProperty(exports, "Constraints", { enumerable: true, get: function () { return types_2.Constraints; } });
Object.defineProperty(exports, "EVENT_NAME_REGEX", { enumerable: true, get: function () { return types_2.EVENT_NAME_REGEX; } });
// Storage implementations (for custom storage adapters)
var storage_1 = require("./storage");
Object.defineProperty(exports, "InMemoryEventStorage", { enumerable: true, get: function () { return storage_1.InMemoryEventStorage; } });
Object.defineProperty(exports, "LocalStorageEventStorage", { enumerable: true, get: function () { return storage_1.LocalStorageEventStorage; } });
Object.defineProperty(exports, "createDefaultStorage", { enumerable: true, get: function () { return storage_1.createDefaultStorage; } });
// Network client (for custom network implementations)
var network_1 = require("./network");
Object.defineProperty(exports, "FetchNetworkClient", { enumerable: true, get: function () { return network_1.FetchNetworkClient; } });
Object.defineProperty(exports, "createDefaultNetworkClient", { enumerable: true, get: function () { return network_1.createDefaultNetworkClient; } });
// Utilities (for advanced usage)
var utils_1 = require("./utils");
Object.defineProperty(exports, "generateAnonymousId", { enumerable: true, get: function () { return utils_1.generateAnonymousId; } });
Object.defineProperty(exports, "generateUUID", { enumerable: true, get: function () { return utils_1.generateUUID; } });
Object.defineProperty(exports, "getISOTimestamp", { enumerable: true, get: function () { return utils_1.getISOTimestamp; } });
Object.defineProperty(exports, "isValidEventName", { enumerable: true, get: function () { return utils_1.isValidEventName; } });
Object.defineProperty(exports, "validateEventName", { enumerable: true, get: function () { return utils_1.validateEventName; } });
Object.defineProperty(exports, "sanitizeProperties", { enumerable: true, get: function () { return utils_1.sanitizeProperties; } });
Object.defineProperty(exports, "detectPlatform", { enumerable: true, get: function () { return utils_1.detectPlatform; } });
Object.defineProperty(exports, "detectDeviceType", { enumerable: true, get: function () { return utils_1.detectDeviceType; } });
Object.defineProperty(exports, "getOSVersion", { enumerable: true, get: function () { return utils_1.getOSVersion; } });
Object.defineProperty(exports, "getDeviceModel", { enumerable: true, get: function () { return utils_1.getDeviceModel; } });
// Logger (for debugging)
var logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
