// @mostly-good-metrics/javascript/dist/cjs/logger.js
"use strict";
/**
 * Internal logger for the MostlyGoodMetrics SDK.
 * Only outputs when debug logging is enabled.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
exports.setDebugLogging = setDebugLogging;
exports.isDebugEnabled = isDebugEnabled;
exports.debug = debug;
exports.info = info;
exports.warn = warn;
exports.error = error;
const LOG_PREFIX = '[MostlyGoodMetrics]';
let debugEnabled = false;
/**
 * Enable or disable debug logging.
 */
function setDebugLogging(enabled) {
    debugEnabled = enabled;
}
/**
 * Check if debug logging is enabled.
 */
function isDebugEnabled() {
    return debugEnabled;
}
/**
 * Log a debug message (only when debug logging is enabled).
 */
function debug(message, ...args) {
    if (debugEnabled) {
        console.log(`${LOG_PREFIX} [DEBUG]`, message, ...args);
    }
}
/**
 * Log an info message (only when debug logging is enabled).
 */
function info(message, ...args) {
    if (debugEnabled) {
        console.info(`${LOG_PREFIX} [INFO]`, message, ...args);
    }
}
/**
 * Log a warning message (always shown).
 */
function warn(message, ...args) {
    console.warn(`${LOG_PREFIX} [WARN]`, message, ...args);
}
/**
 * Log an error message (always shown).
 */
function error(message, ...args) {
    console.error(`${LOG_PREFIX} [ERROR]`, message, ...args);
}
exports.logger = {
    setDebugLogging,
    isDebugEnabled,
    debug,
    info,
    warn,
    error,
};
