// @mostly-good-metrics/javascript/dist/esm/logger.js
/**
 * Internal logger for the MostlyGoodMetrics SDK.
 * Only outputs when debug logging is enabled.
 */
const LOG_PREFIX = '[MostlyGoodMetrics]';
let debugEnabled = false;
/**
 * Enable or disable debug logging.
 */
export function setDebugLogging(enabled) {
    debugEnabled = enabled;
}
/**
 * Check if debug logging is enabled.
 */
export function isDebugEnabled() {
    return debugEnabled;
}
/**
 * Log a debug message (only when debug logging is enabled).
 */
export function debug(message, ...args) {
    if (debugEnabled) {
        console.log(`${LOG_PREFIX} [DEBUG]`, message, ...args);
    }
}
/**
 * Log an info message (only when debug logging is enabled).
 */
export function info(message, ...args) {
    if (debugEnabled) {
        console.info(`${LOG_PREFIX} [INFO]`, message, ...args);
    }
}
/**
 * Log a warning message (always shown).
 */
export function warn(message, ...args) {
    console.warn(`${LOG_PREFIX} [WARN]`, message, ...args);
}
/**
 * Log an error message (always shown).
 */
export function error(message, ...args) {
    console.error(`${LOG_PREFIX} [ERROR]`, message, ...args);
}
export const logger = {
    setDebugLogging,
    isDebugEnabled,
    debug,
    info,
    warn,
    error,
};
