// @mostly-good-metrics/javascript/dist/esm/types.js
/**
 * Custom error class for SDK errors.
 */
export class MGMError extends Error {
    constructor(type, message, options) {
        super(message);
        this.name = 'MGMError';
        this.type = type;
        this.retryAfter = options?.retryAfter;
        this.statusCode = options?.statusCode;
        // Maintains proper stack trace for where error was thrown (V8 engines)
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, MGMError);
        }
    }
}
/**
 * System event names (prefixed with $).
 */
export const SystemEvents = {
    APP_INSTALLED: '$app_installed',
    APP_UPDATED: '$app_updated',
    APP_OPENED: '$app_opened',
    APP_BACKGROUNDED: '$app_backgrounded',
    IDENTIFY: '$identify',
};
/**
 * System property keys (prefixed with $).
 */
export const SystemProperties = {
    DEVICE_TYPE: '$device_type',
    DEVICE_MODEL: '$device_model',
    VERSION: '$version',
    PREVIOUS_VERSION: '$previous_version',
    SDK: '$sdk',
};
/**
 * Default configuration values.
 */
export const DefaultConfiguration = {
    baseURL: 'https://ingest.mostlygoodmetrics.com',
    environment: 'production',
    maxBatchSize: 100,
    flushInterval: 30,
    maxStoredEvents: 10000,
    enableDebugLogging: false,
    trackAppLifecycleEvents: false,
};
/**
 * Validation constraints.
 */
export const Constraints = {
    MAX_EVENT_NAME_LENGTH: 255,
    MAX_BATCH_SIZE: 1000,
    MIN_BATCH_SIZE: 1,
    MIN_FLUSH_INTERVAL: 1,
    MIN_STORED_EVENTS: 100,
    MAX_STRING_PROPERTY_LENGTH: 1000,
    MAX_PROPERTY_DEPTH: 3,
    MAX_PROPERTY_SIZE_BYTES: 10 * 1024, // 10KB
    COMPRESSION_THRESHOLD_BYTES: 1024, // 1KB
};
/**
 * Regular expression for validating event names.
 * Must start with a letter (or $ for system events) followed by alphanumeric and underscores.
 */
export const EVENT_NAME_REGEX = /^\$?[a-zA-Z][a-zA-Z0-9_]*$/;
